crates/ruff_linter/src/rules/ruff/rules/helpers.rs

```diff
@@ -84,3 +87,72 @@ pub(super) fn is_descriptor_class(func: &Expr, semantic: &SemanticModel) -> bool
         })
     })
 }
+
+// fast check to disqualify any string literal that doesn't
+// have opening and closing brackets
+#[inline]
+fn has_brackets(possible_fstring: &str) -> bool {
+    let mut opening = false;
+    let mut closing = false;
+    for c in possible_fstring.chars() {
+        match c {
+            '{' => opening = true,
+            '}' => closing = true,
+            _ => {}
+        }
+    }
+    opening && closing
+}
+
+/// Returns `true` if `source` is valid f-string syntax with qualified, bound variables.
+/// `kwargs` should be the keyword arguments that were passed to function if the string literal is also
+/// being passed to the same function.
+/// If a identifier from `kwargs` is used in `source`'s formatting, this will return `false`,
+/// since it's possible the function could be formatting the literal in question.
+pub(super) fn should_be_fstring(
+    source: &str,
+    kwargs: Option<&Vec<Keyword>>,
+    semantic: &SemanticModel,
+) -> bool {
+    if !has_brackets(source) {
+        return false;
+    }
+
+    let Ok(Expr::FString(ast::ExprFString { value, .. })) =
+        parse_expression(&format!("f\"{source}\""))
```

Comment:
I wonder what happens if you have

```python
a = """b { # comment
c}  d
"""
```

when running in Python 3.12. I would assume that we accept this as a valid f-string :laughing:

Reply:
This works as a valid f-string in Python 3.12:
```
>>> a = f"""b { # comment
... c}  d
... """
>>> a
'b 4  d\n'
```

And yes, we catch this:
```
RUF027.py:37:22: RUF027 [*] `"""b { # comment
    c}  d
    """` may be a formatting string without an `f` prefix
   |
35 |       c = a
36 |       single_line = """ {a} """
37 |       multi_line = a = """b { # comment
   |  ______________________^
38 | |     c}  d
39 | |     """ 
   | |_______^ RUF027
   |
   = help: Add an `f` prefix to the formatting string

ℹ Unsafe fix
34 34 |     a = 4
35 35 |     c = a
36 36 |     single_line = """ {a} """
37    |-    multi_line = a = """b { # comment
   37 |+    multi_line = a = f"""b { # comment
38 38 |     c}  d
39 39 |     """ 
40 40 |     
```

Though we probably want to collapse the multi-line string for the error message somehow.

Reply:
> Though we probably want to collapse the multi-line string for the error message somehow.

I would be in favor of removing it from the message as it seems redundant with `--show-source`.